### ISSUE TITLE
`ipfs-cluster-service -f init` doesn't seem to override config well

### ISSUE THREAD

**Issue by kishansagathiya** [ipfs/ipfs-cluster#532]
#### Pre-check

* [x] This is not a IPFS Cluster website content issue ([file those here](https://github.com/ipfs/ipfs-cluster-website/issues))
* [x] I read the [troubleshooting section](https://cluster.ipfs.io/documentation/troubleshooting/) of the website and it did not help
* [x] I searched for similar issues in the repo without luck
* [x] All my peers are running the same cluster version
* [x] All my peers are configured using the same cluster `secret`

#### Basic information

* [x] Version information (mark as appropiate):
  * [x] Master
  * [ ] Release candidate for next version
  * [ ] Latest stable version
  * [ ] An older version I should not be using
* [x] Type (mark as appropiate):
  * [x] Bug
  * [ ] Feature request
  * [ ] Enhancement
* [x] Operating system (mark as appropiate):
  * [x] Linux
  * [ ] macOS
  * [ ] Windows
  * [ ] Other: `which?`
* [x] Installation method (mark as appropiate):
  * [ ] Binaries from dist.ipfs.io
  * [x] Built from sources
  * [ ] Docker
  * [ ] Snap
  * [ ] Other: `which?`

#### Description
Steps
1. `ipfs-cluster-service init`
2. `ipfs-cluster-service daemon` (Find everything to be working well here)
3. Stop `ipfs-cluster-service daemon`
4. `ipfs-cluster-service -f init` (to override configuration)
5. `ipfs-cluster-service daemon` (Shows errors and eventually the daemon will stop)

Logs here
```
[kishansagathiya@localhost ipfs]$ ipfs-cluster-service -f init
18:05:27.241  INFO     config: Saving configuration config.go:327
ipfs-cluster-service configuration written to /home/kishansagathiya/.ipfs-cluster/service.json
[kishansagathiya@localhost ipfs]$ ipfs-cluster-service daemon
18:05:31.069  INFO    service: Initializing. For verbose output run with "-l debug". Please wait... daemon.go:43
18:05:31.248  INFO    cluster: IPFS Cluster v0.5.0-753322cd listening on:
        /ip4/127.0.0.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/10.215.99.149/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/192.168.42.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/192.168.122.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/172.19.0.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/172.17.0.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/172.18.0.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo

 cluster.go:103
18:05:31.249  INFO    restapi: REST API (HTTP): /ip4/127.0.0.1/tcp/9094 restapi.go:403
18:05:31.249  INFO   ipfshttp: IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/127.0.0.1/tcp/5001 ipfshttp.go:209
18:05:31.249  INFO  consensus: existing Raft state found! raft.InitPeerset will be ignored raft.go:203
18:05:31.250  INFO    restapi: REST API (libp2p-http): ENABLED. Listening on:
        /ip4/127.0.0.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/10.215.99.149/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/192.168.42.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/192.168.122.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/172.19.0.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/172.17.0.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo
        /ip4/172.18.0.1/tcp/9096/ipfs/QmNfSzfyEYKZ54XVW2gPeaZjaEejWzmWwa2NCw3JnEGfbo

 restapi.go:420
18:05:32.772 ERROR       raft: NOTICE: Some RAFT log messages repeat and will only be logged once logging.go:105
18:05:32.773 ERROR       raft: Failed to make RequestVote RPC to {Voter QmVwugPFW471L1xYJYWurh2KzcYvfgXzrM74wpcXzpaLTH QmVwugPFW471L1xYJYWurh2KzcYvfgXzrM74wpcXzpaLTH}: dial attempt failed: failed to dial <peer.ID VwugPF> (default failure) logging.go:105
18:05:33.882 ERROR       raft: Failed to make RequestVote RPC to {Voter QmVwugPFW471L1xYJYWurh2KzcYvfgXzrM74wpcXzpaLTH QmVwugPFW471L1xYJYWurh2KzcYvfgXzrM74wpcXzpaLTH}: dial backoff logging.go:105
18:05:51.250 ERROR    cluster: ***** ipfs-cluster consensus start timed out (tips below) ***** cluster.go:363
18:05:51.250 ERROR    cluster: 
**************************************************
This peer was not able to become part of the cluster.
This might be due to one or several causes:
  - Check the logs above this message for errors
  - Check that there is connectivity to the "peers" multiaddresses
  - Check that all cluster peers are using the same "secret"
  - Check that this peer is reachable on its "listen_multiaddress" by all peers
  - Check that the current cluster is healthy (has a leader). Otherwise make
    sure to start enough peers so that a leader election can happen.
  - Check that the peer(s) you are trying to connect to is running the
    same version of IPFS-cluster.
**************************************************
 cluster.go:364
18:05:51.250  INFO    cluster: shutting down Cluster cluster.go:431
18:05:51.250  INFO  consensus: stopping Consensus component consensus.go:176
18:05:56.259 WARNI  consensus: timed out waiting for state updates before shutdown. Snapshotting may fail raft.go:414
18:05:56.259 ERROR       raft: Failed to take snapshot: nothing new to snapshot logging.go:105
18:05:56.259  INFO    monitor: stopping Monitor pubsubmon.go:154
18:05:56.259  INFO    restapi: stopping Cluster API restapi.go:438
18:05:56.260  INFO   ipfshttp: stopping IPFS Proxy ipfshttp.go:536
18:05:56.260  INFO pintracker: stopping MapPinTracker maptracker.go:119
[kishansagathiya@localhost ipfs]$ ipfs-cluster-service 
```

**Comment by hsanjuan** [comment]
Re-initializing the configuration does not clean up the state and you seem to be using an old one:

`18:05:31.249  INFO  consensus: existing Raft state found! raft.InitPeerset will be ignored raft.go:203`

**Comment by kishansagathiya** [comment]
@hsanjuan 
I see. So `--force` or `-f` only overrides the service.json file and nothing else. Closing this issue.

**Comment by lanzafame** [comment]
I feel like cleaning up the state might end up being a foot gun for someone
down the track. But clearing the peerstore file would be useful.

**Comment by kishansagathiya** [comment]
> I feel like cleaning up the state might end up being a foot gun for someone down the track.

How?

Also, Is there a use case, in which I would just want to override configuration and keep everything else as it is?

**Comment by hsanjuan** [comment]
Perhaps worth just disallowing re-initializing the config, like go-ipfs does

**Comment by hsanjuan** [comment]
@kishansagathiya after sleeping on it I think it's fine to clean up the state when running `-f init` (that is, rotate). This should be fairly easy (look at `ipfs-cluster-service state cleanup`). Wanna take a stab at it?

**Comment by kishansagathiya** [comment]
@hsanjuan 
> Wanna take a stab at it?

Sure

**Comment by kishansagathiya** [comment]
> When you have been messing around with the configuration and broke something and you want to restore it to default.
After updating cluster, and you want to get any new configuration fields by having cluster regenerate the config file.

Huh, right. If we make the proposed modification, we are not providing any way to just restore config back to default without removing state. And anyway there is not real need for this change as we have `ipfs-cluster-service state cleanup` to do this. 
But before I strike down the idea of adding cleanup with force init, I have a question.
How much does the state depend on config? If I change config, would I be able to continue with the state that was generated using a previous config?

**Comment by lanzafame** [comment]
> How much does the state depend on config? If I change config, would I be able to continue with the state that was generated using a previous config?

Yes you should be able to and if you couldn't I would say that there is a bug :)

**Comment by kishansagathiya** [comment]
> Yes you should be able to and if you couldn't I would say that there is a bug :)

In that case, I think we should rather not clean state with `-f`. @hsanjuan WDYT ?

**Comment by hsanjuan** [comment]
> > How much does the state depend on config? If I change config, would I be able to continue with the state that was generated using a previous config?
> 
> Yes you should be able to and if you couldn't I would say that there is a bug :)

The state **depends** on the config. The config has the peer ID and the state stores the peerstore for Raft. If your peer ID changes nothing is going to work with the given state. The secret is also changing so no chance of connecting to your old cluster. I think it makes sense to cleanup the state because `-f init` is suppose to initialize a blank peer, not a peer which won't work. I can however see a bunch of `ipfs-cluster-service config` commands: `config reset`, `config upgrade` that modify the config cleverly (without killing the private key, peer id and cluster secret and api libp2p host and basic auth user list etc).

**Comment by lanzafame** [comment]
> The config has the peer ID and the state stores the peerstore for Raft. If your peer ID changes nothing is going to work with the given state.

Ahh my mistake, I wasn't aware that it did this. 

> I think it makes sense to cleanup the state because -f init is suppose to initialize a blank peer, not a peer which won't work.

I am fine with this if we add a confirmation prompt to make users aware of exactly what will be removed, and also update the usage docs, which currently only state that the configuration will be overwritten.

```sh
...
Note that the --force first-level-flag allows to overwrite an existing
configuration.
...
```

I would love to see `config reset` and `config upgrade` commands added.

**Comment by hsanjuan** [comment]
> I am fine with this if we add a confirmation prompt to make users aware

I disagree, or well, I think that `-f` is the confirmation that the user wants to do it. Reminder that the state is rotated (not deleted). But ok...

Let's remove the global `force` flag and start making local flags for the commands they affect. It is difficult to document them succinctly when using a global flag that applies to several commands. Then we can have `init -f` which will overwrite and rotate (with a prompt) and an additional `-y`/`--yes` flag which will override the prompt.

Note that I hate prompts. I think interactive prompts are the evil in person and everytime I get a prompt on a CLI I want to throw my monitor out of the window. They are only surpassed by interactive ncurses interfaces offering you options used sometimes with no warning by apt-get. But that's just my personal preference. :)

**Comment by lanzafame** [comment]
I will note that I would rather not have a prompt also but I just feel that
`init -f` doesn't  quite convey as much meaning as the action that is about
to occur.

**Comment by hsanjuan** [comment]
@lanzafame agree with the plan above (with prompt) then?

**Comment by kishansagathiya** [comment]
@hsanjuan @lanzafame 
I think it makes more sense with prompt.
Currently `ipfs-cluster-service state cleanup` also prompts before cleaning up

**Comment by lanzafame** [comment]
@hsanjuan yeah that sounds good.

**Comment by kishansagathiya** [comment]
@hsanjuan @lanzafame 
Made required changes. The PR is ready for review https://github.com/ipfs/ipfs-cluster/pull/553#issuecomment-426279322
